perception/autoware_ground_segmentation_cuda/include/autoware/cuda_scan_ground_segmentation/device_vector.hpp

```diff
@@ -0,0 +1,332 @@
+  stream_ = other.stream_;
+  mempool_ = other.mempool_;
+
+  other.data_ = nullptr;
+  other.ele_num_ = other.actual_ele_num_ = 0;
+  other.stream_.reset();
+  other.mempool_.reset();
+}
+
+template <typename T>
+device_vector<T>::device_vector(
+  size_t ele_num, std::shared_ptr<CudaStream> stream, std::shared_ptr<CudaMempool> mem_pool)
+: stream_(stream), mempool_(mem_pool), data_(nullptr)
+{
+  ele_num_ = actual_ele_num_ = 0;
+
+  if (ele_num > 0) {
+    resize(ele_num);
+  }
+}
+
+template <typename T>
+device_vector<T>::device_vector(
+  const std::vector<T> & other, std::shared_ptr<CudaStream> stream,
+  std::shared_ptr<CudaMempool> mem_pool)
+: stream_(stream), mempool_(mem_pool), data_(nullptr)
+{
+  ele_num_ = actual_ele_num_ = 0;
+
+  if (other.empty()) {
+    return;
+  }
+
+  allocate(other.size());
+  copyHtoD(data_, other.data(), other.size());
+  ele_num_ = other.size();
+}
+
+template <typename T>
+device_vector<T> & device_vector<T>::operator=(const device_vector & other)
+{
+  size_t other_size = other.size();
+
+  resize(other_size);
+
+  if (other_size == 0) {
+    copyDtoD(data_, other.data_, other_size);
+  }
+
+  return *this;
+}
+
+template <typename T>
+device_vector<T> & device_vector<T>::operator=(device_vector && other)
+{
+  release();
+
+  data_ = other.data_;
+  ele_num_ = other.ele_num_;
+  actual_ele_num_ = other.actual_ele_num_;
+  stream_ = other.stream_;
+  mempool_ = other.mempool_;
+
+  // TODO: is stream_ = other.stream_ necessary?
+  other.data_ = nullptr;
+  other.ele_num_ = other.actual_ele_num_ = 0;
+  other.mempool_.reset();
+  other.stream_.reset();
+
+  return *this;
+}
+
+template <typename T>
+device_vector<T> & device_vector<T>::operator=(const std::vector<T> & other)
+{
+  release();
+
+  if (other.empty()) {
+    return *this;
+  }
+
+  copyHtoD(data_, other.data(), sizeof(T) * other.size());
+
+  return *this;
+}
+
+template <typename T>
+void device_vector<T>::resize(size_t new_ele_num)
+{
+  // If no significant change, just change the number of elements
+  if (new_ele_num >= actual_ele_num_ * 0.8 && new_ele_num <= actual_ele_num_) {
+    ele_num_ = new_ele_num;
+    return;
+  }
+
+  // Otherwise, reallocate the data
+  release();
+  allocate(new_ele_num);
+  ele_num_ = new_ele_num;
+}
+
+template <typename T>
+void device_vector<T>::reserve(size_t new_ele_num)
+{
+  if (new_ele_num <= actual_ele_num_ * 0.8) {
+    return;
+  }
+
+  allocate(new_ele_num);
+}
+
+template <typename T>
+void device_vector<T>::clear()
+{
+  // Release the memory only
+  release();
+}
+
+template <typename T>
+device_vector<T>::~device_vector()
+{
+  // Release the memory
+  release();
+  // Release the stream and memory pool
+  stream_.reset();
+  mempool_.reset();
+}
+
+template <typename T>
+void device_vector<T>::allocate(size_t ele_num)
+{
+  if (ele_num > 0) {
+    if (mempool_) {
+      CHECK_CUDA_ERROR(
+        cudaMallocFromPoolAsync(&data_, sizeof(T) * ele_num, mempool_->get(), stream_->get()));
```

Comment:
I would recommend zero-clearing the memory region returned by `cudaMallocFromPoolAsync` using `cudaMemsetAsync` because the region may be dirty, and it can cause unexpected behavior.
(Or you can also reuse [pool supported `make_unique` in `autoware::cuda_utils`](https://github.com/autowarefoundation/autoware_universe/blob/cf396aedbe97368a89f8a96dd24b1e6a7b0cdf31/sensing/autoware_cuda_utils/include/autoware/cuda_utils/cuda_unique_ptr.hpp#L59-L79))